### ISSUE TITLE
Clarify ResourceRequirement as hints: "job should not be run" vs best-effort "soft" requirement?

### ISSUE THREAD

**Issue by whlavina** [common-workflow-language/common-workflow-language#877]
Is a `ResourceRequirement` in the `hints` section intended as a best-effort to provision the specified amount of resources? If the answer is no, that hints, if they are supported by the runner, are strict requirements, then how does one specify desired ("soft") requirements?

For example, a process may run best if given, say, 4 cores or 2 GiB of memory, but will adapt to having fewer cores or less memory (perhaps running sub-optimally yet still succeeding); it may also use more cores or memory if provided (perhaps not optimally), up to some maximum value (at which point no further benefit is gained).

The wording in the CWL specification suggests there is no support for such desired ("soft") requirements; the description for `max` does _not_ suggest that the runner will make any attempt to provide the specified amount of resources; rather, it makes a negative assertion that:
> "max" is the maximum amount of a resource that the job shall be permitted to use. [...]  If a job attempts to exceed its "max" resource allocation, an implementation may deny additional resources [...]

Meanwhile, for `min`, the CWL specification for `ResourceRequirement` suggests that failure to provision the resources would be a failure (without clarifying whether as `hints` or as `requitements`):
> If "min" cannot be satisfied, the job should not be run.

See also: Issue #141, which suggests that perhaps `hints` could serve as best-effort indicator:
> CWL also provides "hints" which are similar to requirements except that failure to fulfill a hint is not a failure.
The CWL specification also says the following, also in favor of expressing that `hints` could be a best-effort indicator:
> `hints`: Declares hints applying to either the runtime environment or the workflow engine that may be helpful in executing this process. It is not an error if an implementation cannot satisfy all hints, however the implementation may report a warning.

The reference implementation of `cwltool` suggests that the interpretation is: for `ResourceRequirement` within `hints`, if the `ResourceRequirment` is interpreted by the runner, then it is a strict requirement, and the job will fail if the resources cannot be provisioned. To demonstrate this, run the reference CWL conformance test, `cwltool --parallel steplevel-resreq.cwl`, adjusting the `coresMin` to a high value and changing the `requirement` to `hints`. Despite being `hints`, the test will fail.

If `ResourceRequirement` in `hints` _are_ intended as best-effort provisioning by the runner, then please consider this a bug against the `cwltool` reference runner experimental `-parallel` option.

**Comment by mr-c** [comment]
Hello @whlavina and thank you for your thoughtful issue.

There are indeed some room for improvement in how resources requests are communicated in CWL v1.x.

Right now the model is as follows:

Pick (A) or (B) but not both:

A: Hard minimums and suggested maximums via `ResourceRequirement` under `requirements`
B: Suggested minimums and suggested maximums via `ResourceRequirement` under `hints`

Perhaps a better model would be:

Pick zero or more of the following for each resource type:
1. Hard minimums per resource type
2. Suggested minimums per resource type
3. Suggested maximum per resource type

An even more advanced model could allow for one or more resource type to scale the others based upon its availability. For example: a particular tool scales around number of CPU cores, with disk and memory usage scaling linearly with the number of cores available. However I don't know if there is enough demand for this level of complexity.

**Comment by whlavina** [comment]
To compare with industry job schedulers:
* [Grid Engine/SGE/UGE](https://docs.oracle.com/cd/E19957-01/820-0699/i1003714/index.html): Resource requests are minimums, may be hard vs soft, and are specified in order of importance during job submission. Hard resource requests _must_ be met or the job will not execute; soft resource requests will be met with a best-effort aiming to satisfy the most requests giving precedence to requests specified earlier in request order. There is no mechanism to explicit asserting an optimal target value for a resource.
* [Platform Load Sharing Facility](https://docs.oracle.com/cd/E19957-01/820-0699/i1003714/index.html): Complex scheduling which supports selection (using arbitrary logical and comparative operators), ordering (to prefer among the candidates meeting the selection criteria), and asserting expected resource usage (to inform the scheduler about potentially more optimal allocations).
* [Slurm](https://slurm.schedmd.com/sbatch.html): Complex options for specifying resource requirements as minimums, resource maximums which may constrain job execution (I don't know if exceeding maximums will cause jobs to be terminated), as well as special case options to allow over-subscription of resources.
* [Mesos](http://mesos.apache.org/documentation/latest/reservation/): The Mesos framework supports schedulers with complex resource allocation. Resource reservations ensure the minimums are available for consumption by the job. There is explicit support for over-subscription of resources.
* [Kubernetes](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container): Resources as `limits` and `requests`. Limits are maximums; for CPU, jobs may be slowed but will not be terminated, while for memory, jobs which exceed limits may be terminated. Requests are minimums which _must_ be met or the job will not execute. Note that Kubernetes is a container orchestration platform, not a batch scheduler; its resource scheduler is not very sophisticated.
* [Nomad](https://www.nomadproject.io/docs/job-specification/resources.html): Resources as minimum requirements to execute. There is no mechanism to specify resource limits, expected usage, or optimal resource requests. Note that Nomad is a container orchestration platform, not a batch scheduler; its resource scheduler is not very sophisticated.

It should be noted that some of the above batch schedulers will monitor resource usage to dynamically assess consumption requirements, aiming to obviate the need to explicitly assert expected/maximum resource requirements of existing jobs before considering additional jobs for co-tenancy on a host.

As can be seen from the above inventory, guaranteed minimums for resources are supported by all of these systems. Many of the systems provide mechanisms to specify maximum resource consumption, in some cases this being a best-effort to limit how much a job may use (by throttling or denying resources when requests, e.g. use of `cgroups` to limit available memory), in other cases by terminating any job to attempts to exceed the specified limits. The more advanced batch schedulers allow complex resource allocation, with minumums, maximums, expected usage, and options for over-subscription. The history of batch schedulers is one of shared use of fixed size cluster compute farm, aiming to optimize use of the available resources; the future direction, of Cloud computing, is of a virtually "infinitely scalable" elastic cluster, which is more focused on dynamically growing the compute resources rather than trying to bin-pack jobs optimally with co-tenancy. The latter admits simplifying the model for resource requirements.

In my opinion, it is premature to disregard the use case of a fixed size cluster compute farm. Therefore, I think CWL should go beyond the naive model of merely specifying minimums. I agree that a model supporting hard minimums (guaranteed or else the job can't run), suggested "minimums" or rather target/preferred/optimal/expected resource usage, and suggested maximums (may be throttled/limited) is reasonable. I might also add hard maximum (job will be terminated if exceeded), to stop runaway jobs. Of course, CWL runners might not implement all of these capabilities, with the consequence that workflows might fails if minimums are not honored, or workflows may go awry and over-consume resources if maximums are not honored.

The existing CWL specification effectively supports specifying 4 values, if you consider the `min` vs `max` and `hints` vs `requirements` axes. However, the current descriptions don't match the proposed model, and the current terminology leads to potential misunderstanding:
* `ResourceRequirement`, by having the word _requirement_ in the attribute, leads to confusion when this is specified in `hints`.
* "Suggested minimum", by the presence of the word _minimum_, would obfuscate that this is the target/preferred/optimal/expected resource usage. None of the other job schedulers use the terminology of min/max in the attributes used for specifying resources, which is why the confusion isn't present in those systems.

**Comment by geoffjentry** [comment]
@whlavina I don't think I agree that because many batch schedulers do a thing that a workflow description should do that thing. IMO that's up to the implementation to decide.

In an ideal world, things would look as follows:

- Each tool has a function provided by the **author** describing the cost/performance curve over some resource variable X. There's potentially a minimum value (i.e. "Below this, it'll never work") and there's potentially a maximum value (i.e. "above this and you're just wasting money as there is no performance gain").
- On workflow submission, the **user** describes where on that curve they'd like to live, and what their acceptable range is if it can't be met exactly
- The implementation will either find something within that range or throw an error

As someone who has had a hand in multiple workflow description formats as well as engines, doing the above is a pretty tall task, so we've wound up with a few imperfect not-fully-baked solutions in the world. In the case of CWL it has the min/max as specified, and it's up to the implementation & potentially user to decide. Which might sound like a complete handwave, but that's because it **is** - however that's not a bad thing per se, I go back to there not being a great solution here.

**Comment by whlavina** [comment]
I agree that just because many batch schedulers do a thing, a workflow description should do that thing.

Rather, I was illustrating the current landscape to benefit from the wisdom of experience. Many bright minds have though about the problem of resource allocation and optimizing usage, and it is wise to consider this prior body of work. CWL may support a simpler model, a more complex one, or a different model, to meet its use cases.

I opened this ticket in response to an immediate use case of ours: we have a process which is multi-threaded, and scales mostly linearly with available CPU. The CWL specifications aren't clear about how to express this fact:
* If we specify a very high `minCores` then our workflow will simply fail to run.
* If we specify `minCores=1` then it's not clear that the CWL runner will try to provision more CPU; giving the process a single CPU will cause the wallclock runtime to be significantly elongated.
* If we specify `maxCores` with some arbitrary large number, say `maxCores=1000`, the specifications don't suggest this will make any pragmatic difference; it may as well be a no-op.
* If we specify `maxCores=4` for example, this may artificially constrain our process from using more CPU if it is available, since the specification merely assert use of this requirement _may deny additional resources_.
* Is it expected that single-threaded commands/programs should always assert `maxCores=1`, and that multi-threaded commands/programs are assumed in the absence of this? The specifications say `minCores` has a default value of 1, but no default for `maxCores`. The issue here is of negative vs positive assertions.

**Comment by geoffjentry** [comment]
As a representative of one of the CWL implementations, I feel your pain - I find the status quo to be extremely limiting. The stock answer is that it's up to the implementation to figure it out, but the reality is that since very few **users** of workflows (not just CWL) will override defaults using their own knowledge, that it's a difficult nut to crack. 

However I'd argue that the bright minds of the past were solving for a slightly different problem. So while yes one should (and often does) lean on that wisdom, it's also worth keeping in mind that it can be a bit of an apples & oranges comparison. Even using the various examples available to you at the moment, I'd point out that it's highly dependent on the implementation of CWL - e.g. SLURM does not have that issue.

**Comment by whlavina** [comment]
I agree too, that you're solving a different problem. The use case for the job schedulers, is one where a given compute cluster is concrete and has known resources available, whereas CWL is an abstract specification and must accommodate running on a cluster of unknown capabilities. The desire includes portability, to allow individuals to reuse workflows from others, without having to tweak/edit them.

For example, if you know your compute cluster has machines with 4 CPU cores per host, then you can request 4 cores, and have confidence your jobs will execute at maximum velocity.

So indeed, the CWL use case is perhaps more difficult. I'm happy to discuss anything that can improve the landscape of workflow execution while being pragmatic to implement. If the minimum effort here is merely some clarification of the existing CWL behavior, that may be acceptable.

**Comment by mr-c** [comment]
@whlavina as I mentioned in https://github.com/common-workflow-language/common-workflow-language/issues/877#issuecomment-523349045 I would be open to the idea of a new resource request/limits construct. But we need a concrete proposal, and then an implementation to test out the idea.
